Fully coupling matrix is set as default if a pure Newton is chosen

fdkong opened the pull request:
Looks like users usually forget to use a full coupling matrix
when they are trying to use pure Newton.
Newton depends on the accuracy of Jacobian.
If some entires are ignored automatically, a bad convergence may occur.

Closes #9968

moosebuild commented:
Job [Precheck](https://www.moosebuild.org/job/121124/) on e288289 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format 4f036ed63976d6b55b60b308dccab6baaefdc3f6`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9969/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9969/style.patch | git apply -v`

fdkong committed:
Fully coupling matrix is set as default if a pure Newton is chosen

Looks like users usually forget to use a full coupling matrix
when they are trying to use pure Newton.
Newton depends on the accuracy of Jacobian.
If some entires are ignored automatically, a bad convergence may occur.

Refs #9968

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/121151/) on 7c0f055 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9969/site)

This comment will be updated on new commits.

friedmud reviewed (changes requested):
👎 See my comments here: https://github.com/idaholab/moose/issues/9968#issuecomment-333423456

andrsd commented:
I am with @friedmud on this.

permcody commented:
Hmm, Two strong "no"s from the MOOSE team. This PR was only changing the default, not making it permanent. In lieu of this change, maybe we could consider an info message instead when this combination is used. It wouldn't advocate for a warning, just an info message. I'm fairly neutral on this, but you guys feel pretty strongly about it. This came up because a user just thought he could turn on Newton and go. You may have seen that last week on the mailing list.

andrsd commented:
How about we do not allow `solve_type` in `[Executioner]`? If there is no `[Preconditioning]` block, we do PJFNK with block-diagonal preconditioner, i.e. our current default.  For everything else, users have to use `[Preconditioning]` block?  The parameter `full = true` really makes sense just with SMP and there are cases where NEWTON makes sense, but `full = true` does not.

permcody commented:
> How about we do not allow solve_type in `[Executioner]`? 

That would be a huge change. This is in most input files. Is it a good change? I'm not so sure. Many problems don't require preconditioning blocks so forcing the `solve_type` to appear there is also a little weird.

andrsd commented:
I agree it is counterintuitive to set `solve_type` in `[Preconditioning]` block when one wants to use NEWTON. In attempts to use NEWTON, users do not really think about preconditioning, they want Jacobian. I guess, it was not a good idea...

mangerij commented:
I kind of agree with both sides here. At this point, it's been burned into
my brain to use 'full=true'. But for new users, maybe not so much.

 Our physics (in Ferret), for most boundary conditions absolutely does not
converge unless you use this option. We only have 7 variables (soon 13) so
PJFNK always is a bit overkill (in terms of performance from what ive
tested).

Perhaps just more "alerts" in the documentation to suggest 'full=true' when
using NEWTON is all that is needed. No need to change the default.

On Oct 2, 2017 6:53 PM, "David Andrs" <notifications@github.com> wrote:

> I agree it is counterintuitive to set solve_type in [Preconditioning]
> block when one wants to use NEWTON. In attempts to use NEWTON, users do not
> really think about preconditioning, they want Jacobian. I guess, it was not
> a good idea...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/9969#issuecomment-333594792>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AHPulIFbxhQ4Sk0HVQITvKPSqVAG_FUJks5soRTxgaJpZM4PoASK>
> .
>

lindsayad commented:
Find me the `NEWTON` user who doesn't use `full=true`...and then if you do, I'm going to respond and say MOOSE is supposed to be for multi-physics 😄

friedmud commented:
Well... if you really want to get into it... in my (never very humble)
opinion I don't even like the NEWTON option!  I only ever viewed it as a
"debugging" option for simple, couple-of-physics, problems... it was never
meant to actually be _used_.

It actually drives me kind of crazy that NEWTON, full=true and -pc_type lu
are getting used so heavily by so many people.  MOOSE is designed to help
you move _forward_ about 50 years into the new age of parallel iterative
solvers backed by physics based preconditioning.

But, hey... you do you ;-)

On Mon, Oct 2, 2017 at 7:43 PM Alex Lindsay <notifications@github.com>
wrote:

> Find me the NEWTON user who doesn't use full=true...and then if you do,
> I'm going to respond and say MOOSE is supposed to be for multi-physics 😄
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/9969#issuecomment-333696507>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMQkSFC0tHgiLiCBUiyJejGZ8o_wnks5soXUbgaJpZM4PoASK>
> .
>

lindsayad commented:
I will do me, and I think that's a testament to what you've built and not a detraction. The beauty to me of MOOSE is flexibility and user choice. You talk about how MOOSE was meant to be used to solve problems with millions, tens of millions of degrees of freedom and that's awesome! I've done that; I've used that capability! When I start throwing in postprocessors, non-local user objects, etc...there's no way I'm going to try and go through the effort of trying to code a perfect Jacobian and use NEWTON; I'm going to use PJFNK because I think it's the best tool for the job. If I get OOM errors on Blue Waters using `lu`, then I'm going to use a strong iterative preconditioner because that's going to be the better tool for the job. But if I'm working on a small problem where it's not difficult to code-up a perfect Jacobian, then I'm going to use NEWTON because in my experience it's way faster. And if memory overhead is not a concern and I just want a strong preconditioner, then I'm going to use `lu` and I'm not going to feel a lick of shame for using it. And I **love** this flexibility that MOOSE offers me! I think there's a large space of problems (particularly small problems) that MOOSE is attractive for that perhaps the initial design didn't anticipate, and I think that should be seen as a virtue and not a problem. I would never, ever go back to COMSOL precisely because I don't have 10% of the options and freedom I have with that package as I do with MOOSE. 

MOOSE doesn't always have to be used to hit the biggest research nail. It can be used in simple teaching contexts for introducing Newton-Raphson, teaching *pros* and *cons* of iterative vs direct techniques, traditional Newton-Krylov vs matrix-free Newton-Krylov. The same hammer doesn't work for every nail. Not every user is going to be knowledgeable and/or comfortable with new and modern techniques. I, for one, don't have **any** experience with physics based preconditioning although I would love to learn. In the meantime I'm very happy to use the MOOSE preconditioning techniques I know and that I know work, including direct techniques. I have read some literature about using PBP to efficiently precondition fluid-structure interaction problems with traditional iterative Newton-Krylov that I would be really interested in trying out. That kind of goes back to #8434.

Anyway, if I haven't said it enough, I appreciate the flexibility in MOOSE and I guess I would always err towards trying harder to educate our users as opposed to flatly discouraging options, which I don't think anyone really does. I totally agree that there are a large class of problems where `NEWTON` and `lu` aren't appropriate, but I think there's also some problems where they can be used to great effect. Anyway I'm totally on board with a suggestion like @YaqiWang's where we try to auto-detect what blocks of the Jacobian we need to fill. I would also be on board with just keeping what we have now but flashing a **clearly visible** warning to our users when they try to use NEWTON without `full=true`. I myself have missed plenty of warnings that flash by before the DOF and mesh information are printed for example. Not trying to pick a fight @friedmud!

rcm59 commented:
I'm looking for RUNGE_KUTTA, order = second, dammit

andrsd commented:
> I'm looking for RUNGE_KUTTA, order = second, dammit

[here](https://github.com/idaholab/moose/blob/706debd6d21c152e0d0316ceb9245361d2831d25/framework/include/timeintegrators/ExplicitRK2.h) and [here](https://github.com/idaholab/moose/blob/706debd6d21c152e0d0316ceb9245361d2831d25/framework/src/timeintegrators/ExplicitRK2.C)

friedmud commented:
@lindsayad all valid! 😊

Believe me, I'm glad it's being used for all sorts of stuff!  And I'm definitely glad it's flexible enough to do all the things you're doing with it!

The question is: what should be the default?  To me... the default should reflect the purpose of the project and should be a good "suggestion" about where to start.  Once you've used MOOSE for a little while you're going to start changing things anyway...

permcody commented:
Closing per the 30 day policy. Keep this around Fande - I still think we should take another look at this...
